Using Zips.addEntries with FileSources that refer to directories results in exceptions

I was trying to use Zips to construct a zip file containing an arbitrary set of entries, using FileSource objects.  However, if the FileSource refers to a directory, it results in a "Stream closed" exception:

```Caused by: java.io.IOException: Stream closed

	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at java.io.FilterInputStream.read(FilterInputStream.java:107)
	at org.zeroturnaround.zip.commons.IOUtils.copyLarge(IOUtils.java:322)
	at org.zeroturnaround.zip.commons.IOUtils.copyLarge(IOUtils.java:299)
	at org.zeroturnaround.zip.commons.IOUtils.copy(IOUtils.java:274)
	at org.zeroturnaround.zip.ZipEntryUtil.addEntry(ZipEntryUtil.java:131)
	at org.zeroturnaround.zip.ZipEntryUtil.copyEntry(ZipEntryUtil.java:115)
	at org.zeroturnaround.zip.Zips$CopyingCallback.process(Zips.java:655)
	at org.zeroturnaround.zip.ZipEntryOrInfoAdapter.process(ZipEntryOrInfoAdapter.java:26)
	at org.zeroturnaround.zip.Zips.iterateChangedAndAdded(Zips.java:569)
	... 4 more
```

I think the source of the bug is in this method: `ZipEntryUtil#copyEntry(java.util.zip.ZipEntry, java.io.InputStream, java.util.zip.ZipOutputStream, boolean)`,
where on line 115 it does this: `addEntry(copy, new BufferedInputStream(in), out)`

Instead it should do this: `addEntry(copy, in == null ? null : new BufferedInputStream(in), out)` so the called method does not try to read from the input stream.

--- a/src/main/java/org/zeroturnaround/zip/FileSource.java
+++ b/src/main/java/org/zeroturnaround/zip/FileSource.java
@@ -42,8 +42,13 @@ public class FileSource implements ZipEntrySource {
   }
 
   public ZipEntry getEntry() {
-    ZipEntry entry = ZipEntryUtil.fromFile(path, file);
-    return entry;
+    // A directory must be represented by an entry whose name ends with "/", otherwise
+    // it would be stored as an empty regular file rather than a directory.
+    String name = path;
+    if (file.isDirectory() && !name.endsWith("/")) {
+      name = name + "/";
+    }
+    return ZipEntryUtil.fromFile(name, file);
   }
 
   public InputStream getInputStream() throws IOException {

--- a/src/main/java/org/zeroturnaround/zip/ZipEntryUtil.java
+++ b/src/main/java/org/zeroturnaround/zip/ZipEntryUtil.java
@@ -112,7 +112,10 @@ class ZipEntryUtil {
       copy.setTime(System.currentTimeMillis());
     }
     
-    addEntry(copy, new BufferedInputStream(in), out);
+    // A null stream means the entry has no content (e.g. a directory entry). Keep it
+    // null so addEntry skips the copy; wrapping it in a BufferedInputStream would make
+    // it non-null and the copy would fail with "Stream closed".
+    addEntry(copy, in == null ? null : new BufferedInputStream(in), out);
   }
 
   /**

--- a/src/test/java/org/zeroturnaround/zip/ZipsTest.java
+++ b/src/test/java/org/zeroturnaround/zip/ZipsTest.java
@@ -62,6 +62,69 @@ public class ZipsTest extends TestCase {
     assertTrue(ZipUtil.containsEntry(dest, fileName));
   }
 
+  public void testAddDirectoryEntry() throws IOException {
+    assertDirectorySourceAddsDirectoryEntry("somedir", true);
+  }
+
+  public void testAddDirectoryEntryWithTrailingSlash() throws IOException {
+    assertDirectorySourceAddsDirectoryEntry("somedir/", true);
+  }
+
+  public void testAddEmptyDirectoryEntry() throws IOException {
+    assertDirectorySourceAddsDirectoryEntry("somedir", false);
+  }
+
+  /**
+   * Adds a directory {@link FileSource} under the given path and asserts it lands as the
+   * directory entry "somedir/" (regardless of whether the caller's path had a trailing slash).
+   *
+   * @param sourcePath the path passed to the FileSource
+   * @param withChild whether the directory should contain a child file
+   */
+  private void assertDirectorySourceAddsDirectoryEntry(String sourcePath, boolean withChild) throws IOException {
+    File dir = Files.createTempDirectory("ztzip").toFile();
+    File dest = File.createTempFile("temp", ".zip");
+    try {
+      if (withChild) {
+        new File(dir, "child.txt").createNewFile();
+      }
+
+      Zips.create().destination(dest).addEntry(new FileSource(sourcePath, dir)).process();
+
+      assertTrue("Result zip misses directory entry 'somedir/'", ZipUtil.containsEntry(dest, "somedir/"));
+    }
+    finally {
+      FileUtils.deleteQuietly(dir);
+      FileUtils.deleteQuietly(dest);
+    }
+  }
+
+  public void testAddDirectoryAndFileEntries() throws IOException {
+    File dir = Files.createTempDirectory("ztzip").toFile();
+    File dest = File.createTempFile("temp", ".zip");
+    try {
+      new File(dir, "child.txt").createNewFile();
+
+      // An existing regular file added alongside the directory: a directory source has a
+      // null stream while a file source has real content, so this proves the directory fix
+      // does not break a normal file entry and that both kinds of entry coexist in one zip.
+      String fileName = "TestFile.txt";
+      File file = new File("src/test/resources/" + fileName);
+
+      Zips.create().destination(dest)
+          .addEntry(new FileSource("somedir", dir))
+          .addEntry(new FileSource(fileName, file))
+          .process();
+
+      assertTrue("Result zip misses directory entry 'somedir/'", ZipUtil.containsEntry(dest, "somedir/"));
+      assertTrue("Result zip misses file entry '" + fileName + "'", ZipUtil.containsEntry(dest, fileName));
+    }
+    finally {
+      FileUtils.deleteQuietly(dir);
+      FileUtils.deleteQuietly(dest);
+    }
+  }
+
   public void testRemoveEntry() throws IOException {
     File src = new File(MainExamplesTest.DEMO_ZIP);
 

